Use custom compiled PostgreSQL in Travis for merge commits

byucesoy opened the pull request:
Fixes #1034 

With this commit, we start to use custom compiled PostgreSQL builds in Travis
for merge commits. This allows us to run isolation tests and PostgreSQL's own
regression tests along with our regression tests in Travis.

Since manually compiling PostgreSQL takes more time and we also add new tests,
we only enable running these tests on merge commits.

byucesoy commented:
Tests are failing right now because of #1336 and #1337. Other than that, this PR is ready for review. However it makes sense to review [this PR](https://github.com/citusdata/tools/pull/10) first, because this PR depends on that PR to work properly. After merging that PR, we should change [this line](https://github.com/citusdata/citus/pull/1343/files#diff-354f30a63fb0907d4ad57269548329e3R16) to point correct branch.

anarazel commented on .travis.yml:
```diff
@@ -8,11 +8,13 @@ env:
     - PGVERSION=9.6
 before_install:
-  - git clone -b v0.5.1 --depth 1 https://github.com/citusdata/tools.git
+  - git clone -b support_for_custom_compiled_postgres --depth 1 https://github.com/citusdata/tools.git
```
So, before merging, we'll have to change this back, right? I.e. the changes have to be merged into tools first?

byucesoy replied:
Yes, exactly. I changed this line temporarily to see tests are running as expected. I will change it before merging this PR.

jasonmp85 replied:
Open a PR in `citusdata/tools`, please, as I think we're starting to get quite a few branches open over there and we should sweep them up and make a new tag at some point.

byucesoy replied:
There is a [PR](https://github.com/citusdata/citus/pull/1343) in tools repo. I will change this line once it is merged.

jasonmp85 replied:
Heh that link points back here.

anarazel commented on .travis.yml:
```diff
@@ -8,11 +8,13 @@ env:
     # GitHub API token for citus-bot
     - secure: degV+qb2xHiea7E2dGk/WLvmYjq4ZsBn6ZPko+YhRcNm2GRXRaU3FqMBIecPtsEEFYaL5GwCQq/CgBf9aQxgDQ+t2CrmtGTtI9AGAbVBl//amNeJOoLe6QvrDpSQX5pUxwDLCng8cvoQK7ZxGlNCzDKiu4Ep4DUWgQVpauJkQ9nHjtSMZvUqCoI9h1lBy9Mxh7YFfHPW2PAXCqpV4VlNiIYF84UKdX3MXKLy9Yt0JBSNTWLZFp/fFw2qNwzFvN94rF3ZvFSD7Wp6CIhT6R5/6k6Zx8YQIrjWhgm6OVy1osUA8X7W79h2ISPqKqMNVJkjJ+N8S4xuQU0kfejnQ74Ie/uJiHCmbW5W2TjpL1aU3FQpPsGwR8h0rSeHhJAJzd8Ma ...
+    # Bash command below checks if latest commit is a merge commit, if it is we will use custom compiled PostgreSQL
+    - USE_CUSTOM_PG_BUILD=$(git rev-list --merges HEAD~1..HEAD)
```
If you'd add --count, then this'd always be set to 0 or 1, that seems easier to deal with.

jasonmp85 replied:
This seems post-facto… Travis will build like this _after_ a PR merge, right? Why not use a Travis environment variable to detect whether this is a PR build and build _that_ using this config.

When you push the branch, Travis'll build the HEAD of that branch normally…

When you open a PR, Travis sets certain environment variables and builds a merge commit between the target and subject branches of the PR; we could do that build with the increased scrutiny.

jasonmp85 replied:
The environment variables are defined [here](https://docs.travis-ci.com/user/environment-variables/). `TRAVIS_PULL_REQUEST` might be useful. If it's _false_, run normally, otherwise, do this custom build.

jasonmp85 replied:
The other `TRAVIS_PULL` variables are better, as they are entirely blank when not a merge build, facilitating better use in conditionals.

jasonmp85 commented:
@byucesoy for expediency's sake, I implemented some of the feedback I was going to have about these two reviews, mostly around simplicity. I don't think the build was passing before, but the PostgreSQL 9.6 build appears to work now; unfortunately, there is some sort of `libpq` version problem with the PostgreSQL 9.5 build.

@anarazel the error is `/usr/lib/postgresql/9.6/bin/psql: symbol lookup error: /usr/lib/postgresql/9.6/bin/psql: undefined symbol: PQsetErrorContextVisibility`, which looks like a 9.6 `psql` is being used with a 9.5 `libpq` or something. Any clue what could be happening?

anarazel commented:
On 2017-05-04 22:20:56 -0700, Jason Petersen wrote:
> @anarazel the error is `/usr/lib/postgresql/9.6/bin/psql: symbol lookup error: /usr/lib/postgresql/9.6/bin/psql: undefined symbol: PQsetErrorContextVisibility`, which looks like a 9.6 `psql` is being used with a 9.5 `libpq` or something. Any clue what could be happening?

Hum, that's a weird one.  Did you perchance disable installing the usual
packages we install?  Because we should get a very new libpq from the
system...

jasonmp85 commented:
I worked around it by adding a `apt-get install -y libpq5` after the custom PG install. I believe the `nuke_pg` phase might remove most traces of PostgreSQL from the system, so I don't know what `libpq` we were being left with (is one installed by `make install`?)

Adding that gets the tests passing, but the `hll` fails with [an even weirder error](https://travis-ci.org/citusdata/citus/jobs/228994792#L2579). I have two test variants for "HLL present" and "HLL absent", but this is some sort of in-between state (like the control and extension files are present but it can't find the `so`?)

Testing out skipping the HLL install for now.

jasonmp85 commented:
OK, so there are general problems with the tests themselves (removing the HLL install got me past most of the tests). [This build](https://travis-ci.org/citusdata/citus/jobs/229011007) shows parts of the vanilla tests failing but exits with status zero anyhow. We need the right status to have Travis know about a failure, and without that status, the scripts don't print out the `regression.diffs` file.

Can we verify that, locally, PostgreSQL 9.5 doesn't have problems? And that if it does, `make` exits with a non-zero status?

jasonmp85 commented:
OK, so the problem is now about the `CONTEXT` lines in the 9.5 output. See [here](https://travis-ci.org/citusdata/citus/jobs/229014817#L3322-L3338). I can never remember how to read these diffs, but I'm assuming this has something to do with when PL/pgSQL does or doesn't print CONTEXT changing in 9.6. We've had something similar before; can't remember what caused or fixed it…

byucesoy commented:
There are already PostgreSQL installations(from 9.2 to 9.6) in travis. I guess when we compile PostgreSQL from source, travis uses new binaries but old libraries(In this case new binaries are 9.5 binaries and old libraries are 9.6 libraries).

I added `sudo apt-get --purge -y remove postgresql-*` to nuke_pg to completely remove all existing PostgreSQL artifacts.

Also I modified the $PATH to point correct directory. Before that `which psql` returned `/usr/bin/psql` which is not the location of the binaries we compiled.

jasonmp85 commented:
Thanks for finding a solution; I added back the HLL install and everything looks :100:. Shipping.

jasonmp85 commented:
Oops; I was mistaken. Was looking at the push build, not the pull request one. It has the same failure as before, but now in _both_ 9.6 and 9.5 (before was only 9.5).
